### PR TITLE
Cross validation enhancements

### LINKED ISSUE #80: Neural Net Overfits Data
**Describe the bug**
At the end of an optimization with the neural net learner, the predicted cross sections show signs of overfitting. I haven't done a full study on how this affects optimizations, but it leads to extra local minima in the predicted cost landscape which may make it harder to pick new parameter values. It also makes it harder for the user to interpret the cross sections as they have to guess which features are real and which are due to overfittting.

**Expected behavior**
The predicted landscapes from the neural net learner should fit the data to a reasonable degree; they should be relatively smooth without sharp wiggles and spikes.

**Desktop:**

- OS: Windows 10
- M-LOOP commit bc172f251ef1fb71ca6c2df1f9d82f20f0eb7191 (soon after version 3.1.1)
- TensorFlow 2.1.0

**To Reproduce**
This can be seen by running an optimization with the neural net learner. For convenience I've attached the files from an optimization that demonstrates this behavior: [2020-11-04_1667_raman_cool_n_stage_4.zip](https://github.com/michaelhush/M-LOOP/files/5517314/2020-11-04_1667_raman_cool_n_stage_4.zip). I've also included some example code below to play around with the results in those attached files. I'll omit many of the plots generated but attach enough of them to demonstrate the issue. Generally I'll include one of the cross section plots generated by a single net and the plot that shows the min/max/average of the predictions of the different nets.

First, here is what the predicted cross sections look like at the end of the optimization:
```python
# Set options (Set learner_archive to the correct path for your machine)
learner_archive = 'path/to/learner_archive_2020-11-04_22-53.txt'

# Load in data from archive.
import mloop.visualizations as mlv
learner_visualizer = mlv.create_learner_visualizer_from_archive(learner_archive)

# Show cross sections generated during optimization.
learner_visualizer.do_cross_sections()
```
Which yields four plots, including these two:
image image
There are a lot of sharp edges, wiggles, and local minima which are likely due to overfitting. To check that we can delete the neural nets then recreate them with the same regularization coefficient. The only difference is that their weights will be reinitialized to random values then trained.
```python
# Delete the existing nets and create/train new ones.
for net in learner_visualizer.neural_net:
    net.destroy()
learner_visualizer.create_neural_net()
for j in range(learner_visualizer.num_nets):
    learner_visualizer._fit_neural_net(j)

# Plot new fitted cross sections.
learner_visualizer.do_cross_sections()
```
This yields
image
image
The predicted cross sections are now much smoother, as the cost landscapes presumably are. As a further check we can make sure that training the nets more causes overfitting again. The following section took a while, ~5 minutes, to run on my machine.
```python
n_extra_trainings = 100

for k in range(n_extra_trainings):
    for j in range(learner_visualizer.num_nets):
        learner_visualizer._fit_neural_net(j)

# Plot new fitted cross sections.
learner_visualizer.do_cross_sections()
```
image
image
The nets seem overfit again. For example, the blue curve now has a large spike which is not a real feature of the cost landscape. Running the training routines too many times leads to overfitting, and the training routines are run many times throughout an M-LOOP optimization.

Again I haven't done much looking into how this effects M-LOOP's attempts to optimize. This issue certainly does make it harder for the user to interpret the final cost landscapes though, e.g. to figure out which parameters are important or not. Maybe one solution would be to reinitialize the nets periodically during the optimization?

### PR DESCRIPTION
**Pull request by zakv**
Attempts to fix #80. See that issue for discussion and tests that led to the changes here.

Changes proposed in this pull request:

- Add a user option `update_hyperparameters` for neural net learner optimizations which instructs the learner to tune the regularization coefficient using cross validation.
  - It defaults to `False`, which retains the previous behavior.
- Log the values used for the regularization coefficient and add `NeuralNetVisualizer.plot_regularization_history()` to plot those values.
- Use `loss_raw` instead of `loss_total` in `SingleNeuralNet.cross_validation_loss()`
  - This makes comparison of the loss between nets with different regularization coefficients more fair.
- Create a new net to test the previous regularization coefficient value rather than using the existing net.
  - The existing net has been trained on data that ends up in the cross validation set which gives it a misleadingly low cross validation loss. Creating a new net and only training it on the training data subset puts it on level footing with the nets that try out the other possible regularization coefficient values.
  - This also means that the net is re-initialized every time the hyperparameter tuning is run, even if the regularization coefficient doesn't change. That has the consequence that the early stopping condition has a better chance of preventing overfitting as well, as seen [in this comment](https://github.com/michaelhush/M-LOOP/issues/80#issue-739900705).
- Use a randomized subset of observations for the cross validation set.
- Take the regularization coefficient that leads to the lowest cross validation loss regardless of how small the improvement over the previous regularization coefficient value is.
  - Previously it was only changed if the cross validation loss improved by a factor of 10. However in the tests in #80 the change in cross validation loss between regularization coefficient values that were good and those that led to overfitting/underfitting was often much smaller than a factor of 10.
- Add more values to try for the regularization coefficient when tuning it.
- Re-fit the cost and parameter scalers when the neural net hyperparameter tuning runs.
  - This ensures that the scaled costs and parameters are always ~1 and it takes very little time to run.
- Run the hyperparameter tuning for each net once every 100 runs instead of once every 20 runs.
  - This was done to speed up optimizations as the hyperparameter tuning is fairly slow. [It increased the time](https://github.com/michaelhush/M-LOOP/issues/80#issuecomment-747266315) to do 1000 iterations of a simulated optimization from about 0.5 hours to 5.5 hours when running every 20 runs and trying about 10 different regularization values.
- `SingleNeuralNet._loss()` now returns `loss_reg` as well which can be helpful for debugging.
- Update docstrings and documentation to reflect changes.
- Change 3

When `update_hyperparameters` is set to `False` the behavior is essentially the same as before. The main difference is that the plot from `NeuralNetVisualizer.plot_regularization_history()` will be displayed and show just the initial/constant regularization coefficient value.

### CODE REVIEW

**Comment by zakv** [comment]
As usual, I'm interested in if you have any thoughts on this @charmasaur? I know there are a lot of changes here and a bunch of stuff in #80 though so definitely no pressure to look into this beyond what piques your interest.

**Comment by charmasaur** [comment]
Hey Zak, I'm interested in taking a look but probably won't do so for a few
days, so feel free just to merge anyway, and I can follow up with
comments/PRs if I notice anything.

**Review by charmasaur — COMMENTED** [review]
Looks good, just one question

File: mloop/neuralnet.py [hunk 1]
```diff
@@ -638,43 +654,80 @@ def fit_neural_net(self, all_params, all_costs):
...
+                # Extract the training dataset.
+                train_indices = indices[:split_index]
+                train_params = all_params[train_indices]
+                train_costs = all_costs[train_indices]
+                # Extract the cross validation dataset.
+                cv_indices = indices[split_index:]
+                cv_params = all_params[cv_indices]
+                cv_costs = all_costs[cv_indices]
 
                 orig_cv_loss = self.net.cross_validation_loss(cv_params, cv_costs)
```
[the remark this answers is not in the document]
Comment (zakv) [reply mloop/neuralnet.py:687]: Oh oops yep I forgot to delete that line from the previous version. Thanks for catching that!

Fortunately having it there doesn't change the behavior because `orig_cv_loss` is only used in the following line to set the initial value for `best_cv_loss`, but then the value of `best_cv_loss` is changed to `np.inf` on line 696 before it's used. Still though I'll issue another PR to clean that up.

[the remark this answers is not in the document]
Comment (charmasaur) [reply mloop/neuralnet.py:687]: Oh right, I didn't see that we set it to `inf`.

[the remark this answers is not in the document]
Comment (zakv) [reply mloop/neuralnet.py:687]: Fixed up in #96
